src/lib/message/ArbTxHasher.ts

```diff
@@ -0,0 +1,301 @@
+import { zeroPad } from '@ethersproject/bytes'
+
+/**
+ * Creating an ArbTxType enum for different Arbitrum specific tx types.
+ */
+enum ArbTxType {
+  Deposit = 100,
+  Unsigned = 101,
+  Contract = 102,
+  Retry = 104,
+  SubmitRetryable = 105,
+  Internal = 106,
+}
+
+/**
+ * Creating an interface for each Arbitrum specific tx type.
+ */
+interface DepositTx {
+  type: ArbTxType.Deposit
+  l2ChainId: number
+  fromAddress: string
+  messageNumber: BigNumber
+  destAddress: string
+  l1Value: BigNumber
+}
+
+interface UnsignedTx {
+  type: ArbTxType.Unsigned
+  l2ChainId: number
+  fromAddress: string
+  nonce: number
+  maxFeePerGas: BigNumber
+  gasLimit: BigNumber
+  destAddress: string
+  l1Value: BigNumber
+  data: string
+}
+
+interface ContractTx {
+  type: ArbTxType.Contract
+  l2ChainId: number
+  messageNumber: BigNumber
+  fromAddress: string
+  maxFeePerGas: BigNumber
+  gasLimit: BigNumber
+  destAddress: string
+  l1Value: BigNumber
+  data: string
+}
+
+interface RetryTx {
+  type: ArbTxType.Retry
+  l2ChainId: number
+  nonce: number
+  messageNumber: BigNumber
+  fromAddress: string
+  maxFeePerGas: BigNumber
+  gasLimit: BigNumber
+  destAddress: string
+  l1Value: BigNumber
+  data: string
+  ticketID: string
+  maxRefund: BigNumber
+  submissionFeeRefund: BigNumber
+}
+
+interface SubmitRetryableTx {
+  type: ArbTxType.SubmitRetryable
+  l2ChainId: number
+  fromAddress: string
+  messageNumber: BigNumber
+  l1BaseFee: BigNumber
+  destAddress: string
+  l2CallValue: BigNumber
+  l1Value: BigNumber
+  maxSubmissionFee: BigNumber
+  excessFeeRefundAddress: string
+  callValueRefundAddress: string
+  gasLimit: BigNumber
+  maxFeePerGas: BigNumber
+  data: string
+}
+
+interface InternalTx {
+  type: ArbTxType.Internal
+  l2ChainId: number
+  data: string
+}
+
+export class ArbTxHasher {
+  /**
+   * Parse the input data into Uint8Array format
+   * @param BigNumber
+   * @returns Uint8Array
+   */
+  private static formatNumber(value: BigNumber): Uint8Array {
+    return ethers.utils.stripZeros(value.toHexString())
+  }
+
+  /**
+   * Return the hex encoding of Arbitrum specific tx types.
+   * @param ArbTxType
+   * @returns Hex encoding of ArbTxType
+   */
+  private static arbTxTypeToHex(txType: ArbTxType) {
+    switch (txType) {
+      case ArbTxType.Deposit:
+        return '0x64'
+      case ArbTxType.Unsigned:
+        return '0x65'
+      case ArbTxType.Contract:
+        return '0x66'
+      case ArbTxType.Retry:
+        return '0x68'
+      case ArbTxType.SubmitRetryable:
+        return '0x69'
+      case ArbTxType.Internal:
+        return '0x6A'
+    }
+  }
+
+  /**
+   * Accept the Arbitrum specific tx type and call the right function for creating the custom hash.
+   * @param ArbTxType
+   * @returns
+   */
+  public static hash(
+    tx:
+      | DepositTx
+      | UnsignedTx
+      | ContractTx
+      | RetryTx
+      | SubmitRetryableTx
+      | InternalTx
```

Comment:
I think you can have each of the type extend a base type (e.g. with the type field only) so this can be cleaner.